### DISCUSSION TITLE
Rasterize a pdf to another pdf

### DISCUSSION THREAD

**Discussion by maxux** [pymupdf/PyMuPDF#1183] [Ideas]
I'm starting from the idea you can « censor » part of the document by adding an overlay, but like wiki said, it's just an overlay, text is still present on the document.

I'm looking for a way where you can't get original content. One easy idea is to simply rasterize content, so original text won't be physically present at all. This is clearly not optimal for document size, but it works :p 

There is already a `page.get_pixmap()` which does that actually. The feature request here would be a simple way to convert a whole document to a rasterized version of that document.

In the meantime, I'm testing some snippet to do it manually. :)

**Comment by JorjMcKie** [comment]
> The feature request here would be a simple way to convert a whole document to a rasterized version of that document.

You mean a feature for these 7-10 lines?
```python
mat = fitz.Matrix(zoom, zoom)  # zoom = parameter
source = fitz.open(...)
target = fitz.open()
for page in source:
    pix = page.get_pixmap(matrix=mat)
    tarpage = target.new_page(width=pix.width, height=pix.height)
    tarpage.insert_image(tarpage.rect, stream=pix.pil_tobytes("JPEG"))
target.ez_save(targetname)  # targetname = parameter
```
How about you write a `Document` method covering this and submit a Pull Request 😎?

**Reply by maxux** [reply]
I tried and it works quite well ! I'm checking for the PR :)

**Reply by JorjMcKie** [reply]
Just write a function similar to my suggestion, that contains all parameters you want. First - positional - parameter must be the document. I will then incorporate it.

**Reply by JorjMcKie** [reply]
Especially consider whether you want information from the `doc.metadata` partially copied over. I could image that a reference to the source file would be an idea ...

**Reply by JorjMcKie** [reply]
One of the parameters should definitely be, whether annotations should be contained in the pages' pixmaps.

This method would be an extreme version of `doc.scrub()`: Please check the docu for background.

**Reply by JorjMcKie** [reply]
I noticed I made an error in my example. This line:
```python
tarpage = target.new_page(width=pix.width, height=pix.height)
```
should be this:
```python
tarpage = target.new_page(width=page.rect.width, height=page.rect.height)
```
The output PDF should have same page dimensions - independently from the `zoom` factor! A higher zoom would not increase target page dimensions, but instead only improve the image resolution.

**Reply by PasaOpasen** [reply]
```python

from typing import Union, Sequence
import collections.abc
import fitz

def rastr_pdf(
    input_path: str,
    output_path: str,

    dpi: Union[float, Sequence[float]]
):

    with fitz.open(input_path) as doc:
        page_count = doc.page_count

        if not isinstance(dpi, collections.abc.Sequence):
            dpi = [dpi] * page_count
        else:
            assert len(dpi) == page_count

        with fitz.open() as target:

            for page, _dpi in zip(doc, dpi):
                zoom = _dpi / 72
                mat = fitz.Matrix(zoom, zoom)
                pix = page.get_pixmap(matrix=mat)
                tarpage = target.new_page(width=page.rect.width, height=page.rect.height)
                tarpage.insert_image(tarpage.rect, stream=pix.pil_tobytes("JPEG"))

            target.ez_save(output_path)

```
